model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java

```diff
@@ -0,0 +1,204 @@
+     * is committed. If such a constraint violation is detected at the time of the transaction commit, it should
+     * throw an exception and the transaction should roll back.
+     * <p />
+     * The JPA Map Store implements this with a unique index, which is checked by the database both at the time of
+     * insertion and at the time the transaction is committed.
+     */
+    private MapLockEntity lock(String lockName) {
+        prepareTx();
+        DefaultModelCriteria<MapLockEntity> mcb = criteria();
+        mcb = mcb.compare(MapLockEntity.SearchableFields.NAME, ModelCriteriaBuilder.Operator.EQ, lockName);
+        Optional<MapLockEntity> entry = tx.read(QueryParameters.withCriteria(mcb)).findFirst();
+
+        if (entry.isEmpty()) {
+            MapLockEntity entity = DeepCloner.DUMB_CLONER.newInstance(MapLockEntity.class);
+            entity.setName(lockName);
+            entity.setKeycloakInstanceIdentifier(getKeycloakInstanceIdentifier());
+            entity.setTimeAcquired(Time.currentTimeMillis());
+            return tx.create(entity);
+        } else {
+            throw new LockAcquiringTimeoutException(lockName, entry.get().getKeycloakInstanceIdentifier(), Instant.ofEpochMilli(entry.get().getTimeAcquired()));
+        }
+    }
+
+    /**
+     * Unlock the previously created lock.
+     * Will fail if the lock doesn't exist, or has a different owner.
+     */
+    private void unlock(MapLockEntity lockEntity) {
+        prepareTx();
+        MapLockEntity readLockEntity = tx.read(lockEntity.getId());
+
+        if (readLockEntity == null) {
+            throw new RuntimeException("didn't find lock - someone else unlocked it?");
+        } else if (!lockEntity.isLockUnchanged(readLockEntity)) {
+            // this case is there for stores which might re-use IDs or derive it from the name of the entity (like the file store map store does in some cases).
+            throw new RuntimeException(String.format("Lock owned by different instance: Lock [%s] acquired by keycloak instance [%s] at the time [%s]",
+                    readLockEntity.getName(), readLockEntity.getKeycloakInstanceIdentifier(), readLockEntity.getTimeAcquired()));
+        } else {
+            tx.delete(readLockEntity.getId());
+        }
+    }
+
+    private void releaseAllLocks() {
+        prepareTx();
+        DefaultModelCriteria<MapLockEntity> mcb = criteria();
+        tx.delete(QueryParameters.withCriteria(mcb));
+    }
+
+    private static String getKeycloakInstanceIdentifier() {
+        long pid = ProcessHandle.current().pid();
+        String hostname;
+        try {
+            hostname = InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            hostname = "unknown-host";
+        }
+
+        String threadName = Thread.currentThread().getName();
+        return threadName + "#" + pid + "@" + hostname;
```

Comment:
@mhajas - I still think it is a good way to create a meaningful unique name for the instance name as it will not only be instances fighting for locks with other instances, but also threads fighting for locks with other threads. If someone would do debgging on the instance identified with the host name, they could create a thread dump and point to the thread with the lock by using this information. 

Therefore I'd like to keep the thread name.